src/main.cpp

```diff
@@ -3305,9 +3391,17 @@ static bool IsSuperMajority(int minVersion, const CBlockIndex* pstart, unsigned
 
 bool ProcessNewBlock(CValidationState &state, CNode* pfrom, CBlock* pblock, bool fForceProcessing, CDiskBlockPos *dbp)
 {
+    int height = 0;
+    {
+        LOCK(cs_main);
+        // FIXME: IS THIS THE RIGHT HEIGHT TO USE?
+        height = chainActive.Height() + 1;
```

Comment:
Ooh, this is a problem. I don't believe we can assume that the block being processed would end up at the chain tip; we don't know its height until after `AcceptBlock` is called below. This means that we cannot check any consensus rules in `CheckBlock` that require knowledge of the block's height, and thus we can never pass `consensusBranchId` in. We should therefore remove `consensusBranchId` as an argument to `CheckBlock`, `CheckTransaction` and `CheckTransactionWithoutProofVerification`.

This affects two rules:

- https://github.com/zcash/zcash/pull/2925/files#r166610870 which I was already on the fence about, so that means we just need to move that rule out of `CheckTransaction` for now, and can move it back after Overwinter.
- The `joinSplitSig` check in `CheckTransaction`, which in #2903 needs to know `consensusBranchId` from Overwinter.

The one annoyance is that there is currently no common place where general contextual checks are made on transactions, so we will need to move them into both `AcceptToMemoryPool` and `ContextualCheckBlock`. It might make sense to add a new function to contain them, assuming that we want the same DoS level for both mempool and blockchain. Upstream AFAIR hasn't refactored these kinds of rules yet, probably for this reason.

Reply:
I've just grepped around, and `ContextualCheckBlock` is called from `AcceptBlock` (i.e. `ProcessNewBlock`) and `TestBlockValidity`, so that means we should still always validate the above rules when a new block is received. It *isn't* called from `VerifyDB` (only `CheckBlock` and `ContextualCheckInputs` are triggered from there), but that should be fine, as it's only a consistency check.